madminer/utils/morphing.py

```diff
@@ -396,33 +604,61 @@ def calculate_morphing_weights(self, theta, basis=None, morphing_matrix=None):
             raise RuntimeError(
                 "No components defined. Use morpher.set_components() or morpher.find_components() " "first!"
             )
+        if self.gd is None and self.gp is None and self.gs is None:
+            if basis is None:
+                basis = self.basis
+                morphing_matrix = self.morphing_matrix
+
+            if basis is None:
+                raise RuntimeError(
+                    "No basis defined or given. Use PhysicsMorpher.set_basis(), PhysicsMorpher.optimize_basis(), or the "
+                    "basis keyword."
+                )
 
-        if basis is None:
-            basis = self.basis
-            morphing_matrix = self.morphing_matrix
+            if morphing_matrix is None:
+                morphing_matrix = self.calculate_morphing_matrix(basis)
 
-        if basis is None:
-            raise RuntimeError(
-                "No basis defined or given. Use PhysicsMorpher.set_basis(), PhysicsMorpher.optimize_basis(), or the "
-                "basis keyword."
-            )
+            # Calculate component weights
+            component_weights = np.zeros(self.n_components)
+            for c in range(self.n_components):
+                factor = 1.0
+                for p in range(self.n_parameters):
+                    factor *= float(theta[p] ** self.components[c, p])
+                component_weights[c] = factor
+            component_weights = np.array(component_weights)
 
-        if morphing_matrix is None:
-            morphing_matrix = self.calculate_morphing_matrix(basis)
+            # Transform to basis weights
+            return morphing_matrix.T.dot(component_weights)
+        else:
+            if gs is None:
+                gs = self.gs
+            if gd is None:
+                gd = self.gd
+            if gp is None:
+                gp = self.gp
+            
+            if gs is None and gd is None and gp is None:
+                raise RuntimeError(
+                    "No basis defined or given. Use PhysicsMorpher.set_basis(), PhysicsMorpher.optimize_basis(), or the "
+                    "basis keyword."
+                )
+            
+            if morphing_matrix is None:
+                morphing_matrix = self.calculate_morphing_matrix(basis, gs, gd, gp)
 
-        # Calculate component weights
-        component_weights = np.zeros(self.n_components)
-        for c in range(self.n_components):
-            factor = 1.0
-            for p in range(self.n_parameters):
-                factor *= float(theta[p] ** self.components[c, p])
-            component_weights[c] = factor
-        component_weights = np.array(component_weights)
+            component_weights = np.zeros(self.n_components)
+            for c in range(self.n_components):
+                factor = 1.0
+                for p in range(self.n_parameters):
+                    factor *= float(theta[p] ** self.components[c, p])
+                component_weights[c] = factor
```

Comment:
This is occurrence number 2 of the _component weights calculation_ snippet.

Please extract these lines to a private method within this class.